packages/shared/src/utils/deprecated.ts

```diff
@@ -0,0 +1,61 @@
+/**
+ * Mark class methods or functions as deprecated.
+ *
+ * A console WARNING will be displayed when class methods
+ * or functions are invoked.
+ *
+ * Examples
+ * 1. Deprecate class method
+ * class Example {
+ *   getSomething = (arg1, arg2) => {
+ *       deprecated('Example.getSomething', 'Use `getSomethingElse` instead.');
+ *       return `getSomethingValue:${arg1 || '-'}:${arg2 || '-'}`;
+ *   };
+ * }
+ *
+ * 2. Deprecate function
+ * const getSomething = () => {
+ *   deprecated('getSomething', 'Use `getSomethingElse` instead.');
+ *   return 'getSomethingValue';
+ * };
+ */
+export const deprecated = (fnName: string, warning: string): void => {
+  console.warn(`${fnName} is deprecated and will be removed in the next major release.\n${warning}`);
```

Comment:
Can we restrict these warnings to development only? They aren't actionable in a production environment and will likely cause a lot of log pollution and user frustration.

Also, I understand wanting to be noisy here, but I think warning on every function call might be too much. That's potentially a lot of noise if we're calling this in a commonly used method. Users might not be able to upgrade immediately for some reason, and with no way to turn this off it could be frustrating.

Adding a module-level check seems like a reasonable approach, that way it would only warn once per session if in the browser, or once per process otherwise.

Reply:
>Can we restrict these warnings to development only? They aren't actionable in a production environment and will likely cause a lot of log pollution and user frustration.

Yes, we restrict those warnings to be visible in development mode using https://github.com/clerkinc/javascript/pull/1752

> Also, I understand wanting to be noisy here, but I think warning on every function call might be too much. That's potentially a lot of noise if we're calling this in a commonly used method. Users might not be able to upgrade immediately for some reason, and with no way to turn this off it could be frustrating.

Are you suggesting we make it called once?

> Adding a module-level check seems like a reasonable approach, that way it would only warn once per session if in the browser, or once per process otherwise.

Could you elaborate more on this? Could you provide a usage example?

Reply:
> Are you suggesting we make it called once?

Yes, I'd suggest that, too.

For example, if `functionX` is used by the user in multiple places, it only makes sense to throw this warning the first time `functionX` is used. Then people know what they need to switch. All other occurrences wouldn't give any additional benefit to the user (and clutter the console).

> Could you elaborate more on this? Could you provide a usage example?

Instead of checking inside the `deprecate` function if it was already called, you move the scope to the **module scope**.

An example would be:

```ts
const displayedWarnings = new Set<string>()

export const reportOnce = (
  message: string,
  key?: string
): void => {
  const messageId = key ?? message
  if (!displayedWarnings.has(messageId) && !isDevelopmentMode) {
    console.warn(message)
  }
}
```

Reply:
@LekoArts I will proceed with your suggestion but I would like to add here that due to HMR in Remix and Nextjs the modules are reloaded in every request disabling the caching of information using module-level scope.

Relevant: https://www.prisma.io/docs/guides/other/troubleshooting-orm/help-articles/nextjs-prisma-client-dev-practices

PS: You can check the JS-164 internal ticket.